File: src/leetcode/python/medium/CourseSchedule_207.py
class CourseSchedule_207(object):
    """
    if node v has not been visited, then mark it as 0.
    if node v is being visited, then mark it as -1. If we find a vertex marked as -1 in DFS, then their is a ring.
    if node v has been visited, then mark it as 1. If a vertex was marked as 1, then no ring contains v or its successors.
    """

    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        def dfs(courses):
            if visited[courses] == -1:
                return False
            if visited[courses] == 1:
                return True
            visited[courses] = -1
            for j in graph[courses]:
                if not dfs(j):
                    return False
            visited[courses] = 1
            return True

        if len(prerequisites) == 0:
            return True
        visited = [0 for _ in range(numCourses)]
        graph = [[] for _ in range(numCourses)]
        for x, y in prerequisites:
            graph[x].append(y)
        for i in range(numCourses):
            if not dfs(i):
                return False
        return True

File: src/leetcode/java/medium/CourseSchedule_207.java
package leetcode.java.medium;

public class CourseSchedule_207 {
  public boolean canFinish(int numCourses, int[][] prerequisites) {
    if (prerequisites.length == 0) return true;

    int[] visited = new int[numCourses];
    int[] onStack = new int[numCourses];
    for (int i = 0; i < numCourses; i++) {
      if (!dfs(i, prerequisites, visited, onStack)) return false;
    }
    return true;
  }

  private boolean dfs(int i, int[][] prerequisites, int[] visited, int[] onStack) {
    if (visited[i] == 0) {
      visited[i] = 1;
      onStack[i] = 1;
      for (int j = 0; j < prerequisites.length; j++) {
        if (prerequisites[j][1] == i) {
          if (!dfs(prerequisites[j][0], prerequisites, visited, onStack)) return false;
        }
      }
      onStack[i] = 0;
    } else if (onStack[i] == 1) {
      return false;
    }
    return true;
  }
}
